### ISSUE TITLE
actor system does not shut down gracefully

### ISSUE THREAD

**Issue by nimbusgo** [ibm-et/spark-kernel#194]
When you call shutdown on KernelBootstrap there are threads that remain running and prevent the process from ending. 

![screen shot 2015-11-04 at 3 51 37 pm](https://cloud.githubusercontent.com/assets/2001660/10955890/f738aca8-830b-11e5-93da-6828aeb13718.png)

here's a thread dump of the 2 categories that seem to be lingering: 
"iothread-2@7680" prio=5 tid=0x8d nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at sun.nio.ch.KQueueArrayWrapper.kevent0(KQueueArrayWrapper.java:-1)
	  at sun.nio.ch.KQueueArrayWrapper.poll(KQueueArrayWrapper.java:200)
	  at sun.nio.ch.KQueueSelectorImpl.doSelect(KQueueSelectorImpl.java:103)
	  at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:87)
	  - locked <0x2642> (a sun.nio.ch.KQueueSelectorImpl)
	  - locked <0x2643> (a java.util.Collections$UnmodifiableSet)
	  - locked <0x2644> (a sun.nio.ch.Util$2)
	  at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:98)
	  at zmq.Poller.run(Poller.java:207)
	  at java.lang.Thread.run(Thread.java:745)

"reaper-1@7679" prio=5 tid=0x8c nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at sun.nio.ch.KQueueArrayWrapper.kevent0(KQueueArrayWrapper.java:-1)
	  at sun.nio.ch.KQueueArrayWrapper.poll(KQueueArrayWrapper.java:200)
	  at sun.nio.ch.KQueueSelectorImpl.doSelect(KQueueSelectorImpl.java:103)
	  at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:87)
	  - locked <0x2645> (a sun.nio.ch.KQueueSelectorImpl)
	  - locked <0x2646> (a java.util.Collections$UnmodifiableSet)
	  - locked <0x2647> (a sun.nio.ch.Util$2)
	  at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:98)
	  at zmq.Poller.run(Poller.java:207)
	  at java.lang.Thread.run(Thread.java:745)

I suppose it's possible some zmq socket is not getting closed properly on shutdown. Has anyone been able to successfully shutdown the kernel before and actually end all the related threads? I'm currently experimenting with running more than one kernel in the same jvm and these lingering threads are really annoying.

**Comment by lbustelo** [comment]
We probably don't see these issues because we let the notebook manage the process and it ends up doing a kill -9. Regardless, we need to investigate this.

**Comment by jodersky** [comment]
@Lull3rSkat3r and I will have a go at this

**Comment by jodersky** [comment]
Sorry about the long radio silence. On Friday I managed to narrow the bug down to a single test. Thinking I had diagnosed the issue, I tried running the tests with and without forking and with and without parallelization and I got freezes at different places each time!

So I'm back to the start, all I can say is that this looks like some nasty issue involving deadlocks somewhere in socket-related code :(
